packages/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts

```diff
@@ -91,9 +91,10 @@ export function useGridDimensions(
   const pinnedColumns = useGridSelector(apiRef, gridVisiblePinnedColumnDefinitionsSelector);
   const densityFactor = useGridSelector(apiRef, gridDensityFactorSelector);
   const rowHeight = Math.floor(props.rowHeight * densityFactor);
-  const headerHeight = Math.floor(props.columnHeaderHeight * densityFactor);
+  const headerDensityFactor = useGridSelector(apiRef, gridHeaderDensityFactorSelector);
+  const headerHeight = Math.floor(props.columnHeaderHeight * headerDensityFactor);
   const headerFilterHeight = Math.floor(
-    (props.headerFilterHeight ?? props.columnHeaderHeight) * densityFactor,
+    (props.headerFilterHeight ?? props.columnHeaderHeight) * headerDensityFactor,
```

Comment:
We have a separate issue to deal with the fact that compact header filters are overflowing; this change neither fixes or worsens this issue https://github.com/mui/mui-x/issues/13048